In [7]:
from services.db.supabase_services import supabase_client
supabase = supabase_client()

page_size = 1000  # Maximum page size for Supabase
offset = 0
all_results = []

while True:
    results = (
        supabase.table("agents")
        .select('*', count='exact')
        .eq('userId', "user_2mmXezcGmjZCf88gT2v2waCBsXv")
        .range(offset, offset + page_size - 1)
        .execute()
    )
    
    all_results.extend(results.data)
    
    # If we got less results than page_size, we've reached the end
    if len(results.data) < page_size:
        break
        
    offset += page_size

results = all_results  # Store all results in the original variable name

In [4]:
results

[{'created_at': '2024-09-30T09:50:24.660528+00:00',
  'header': '## Title: Event Planning - professional event management & total event solutions ## Description: Peri Peri provides complete event planning solutions - specialising in bespoke party planning, marquees, wedding planning & corporate events',
  'token_count': 54,
  'url': 'https://www.periperi.co.uk',
  'user_id': 'user_2mmXezcGmjZCf88gT2v2waCBsXv',
  'user_name': None,
  'tag': None,
  'root_url': 'https://www.periperi.co.uk',
  'id': '203bb571-e652-4130-a692-d552984bd50a',
  'rag_query': None},
 {'created_at': '2024-09-30T09:50:59.87365+00:00',
  'header': '## Title: Party Organisers- Event management and party planners: contact details ## Description: Party organisers - complete list of contact details, including direct lines for different event departments and emergency 24hour contact details',
  'token_count': 490,
  'url': 'https://www.periperi.co.uk/party-organisers',
  'user_id': 'user_2mmXezcGmjZCf88gT2v2waCBsXv',
 

In [12]:
import requests

url = "http://localhost:8000/api/v1/stripe/create-payment-link"
payload = {
    "product_id": "prod_RcfvpRgzSUvVXj",
    "quantity": 1,
    "unit_amount": 249,
    "currency": "usd"  # optional, defaults to "usd"
}

response = requests.post(url, json=payload)
print(response.json())

{'payment_link': 'https://buy.stripe.com/cN29Bh3encMc2PK8wJ'}


In [10]:
from services.twilio import get_available_numbers
from typing import Dict, Any
from services.twilio import client

country_code = "GB"

#get_available_numbers("HK")
# Try to list up to 5 numbers of each type
    # Map our internal types to Twilio's pricing types
number_type_mapping = {
    'local': 'local',
    'toll_free': 'toll free',
    'mobile': 'mobile',
    'national': 'national'
}
number_types = list(number_type_mapping.keys())
available_numbers: Dict[str, Dict] = {}
monthly_cost: Dict[str, float] = {}

for number_type in number_types:
    try:
        print(f"Getting {number_type} numbers for {country_code}")
        # Try to list up to 5 numbers of each type
        numbers = getattr(client.available_phone_numbers(country_code), number_type).list(limit=5)
        numbers_list = [number.phone_number for number in numbers]

        country_pricing = client.pricing.v1.phone_numbers.countries(country_code).fetch()
        country_pricing = country_pricing.phone_number_prices

        print(f"Got {len(numbers_list)} {number_type}, numbers {numbers_list}")
        # print(f"Pricing: {country_pricing}")
    except Exception as e:
        print(f"Error getting {number_type} numbers for {country_code}: {str(e)}")
        continue

country_pricing

Getting local numbers for GB
Got 5 local, numbers ['+441479787912', '+441626248595', '+441928243483', '+441218275213', '+441174638214']
Getting toll_free numbers for GB
Got 5 toll_free, numbers ['+448081607334', '+448081607421', '+448085012741', '+448081607239', '+448081607311']
Getting mobile numbers for GB
Got 5 mobile, numbers ['+447400479605', '+447401483422', '+447400461332', '+447400481135', '+447400430858']
Getting national numbers for GB
Error getting national numbers for GB: ('Unable to fetch page', HTTP 404 {"code": 20404, "message": "The requested resource /2010-04-01/Accounts/AC2e61d0942c25dd6db0c6c1049fe1d3b1/AvailablePhoneNumbers/GB/National.json was not found", "more_info": "https://www.twilio.com/docs/errors/20404", "status": 404})


[{'number_type': 'local',
  'base_price': '0.918735',
  'current_price': '0.918735'},
 {'number_type': 'mobile',
  'base_price': '0.918735',
  'current_price': '0.918735'},
 {'number_type': 'national',
  'base_price': '0.918735',
  'current_price': '0.918735'},
 {'number_type': 'toll free',
  'base_price': '1.717635',
  'current_price': '1.717635'}]

In [4]:
def purchase_phone_number(phone_number: str) -> Dict[str, Any]:
    """
    Purchase a specific phone number from Twilio
    
    Args:
        phone_number (str): The phone number to purchase in E.164 format (e.g., '+1234567890')
    
    Returns:
        Dict[str, Any]: Dictionary containing the purchased number details
        
    Raises:
        Exception: If the purchase fails
    """
    try:
        purchased_number = client.incoming_phone_numbers.create(
            phone_number=phone_number
        )
        return {
            'phone_number': purchased_number.phone_number,
            'sid': purchased_number.sid,
            'friendly_name': purchased_number.friendly_name,
            'capabilities': {
                'voice': purchased_number.capabilities.get('voice', False),
                'sms': purchased_number.capabilities.get('sms', False),
                'mms': purchased_number.capabilities.get('mms', False)
            }
        }
    except Exception as e:
        raise Exception(f"Failed to purchase number: {str(e)}")



response = purchase_phone_number("+13614281772")
response


{'phone_number': '+13614281772',
 'sid': 'PN0e598cf7f7c8da7b7d57823cc334bf31',
 'friendly_name': '(361) 428-1772',
 'capabilities': {'voice': True, 'sms': True, 'mms': True}}

In [33]:
def get_pricing_for_numbers(client, destination_numbers, origination_number=None):
    results = []
    for number in destination_numbers:
        try:
            pricing = client.pricing.v2.numbers(number).fetch(
                origination_number=origination_number
            )
            results.append({
                'number': number,
                'country': pricing.country,
                'price_unit': pricing.price_unit,
                'terminating_prefix_prices': pricing.terminating_prefix_prices,
                'originating_call_price': pricing.originating_call_price
            })
        except Exception as e:
            results.append({
                'number': number,
                'error': str(e)
            })
    return results

# Usage
pricing_results = get_pricing_for_numbers(client, numbers['local'], '+0987654321')

pricing_results

[{'number': '+16282598072',
  'country': 'United States',
  'price_unit': 'GBP',
  'terminating_prefix_prices': [{'destination_prefixes': ['1628'],
    'origination_prefixes': ['ALL'],
    'friendly_name': 'Trunking Outbound Minute - United States - 48 States (Zone 1)',
    'base_price': None,
    'current_price': '0.004234'}],
  'originating_call_price': {'number_type': 'local',
   'base_price': None,
   'current_price': '0.002716'}},
 {'number': '+16282598875',
  'country': 'United States',
  'price_unit': 'GBP',
  'terminating_prefix_prices': [{'destination_prefixes': ['1628'],
    'origination_prefixes': ['ALL'],
    'friendly_name': 'Trunking Outbound Minute - United States - 48 States (Zone 1)',
    'base_price': None,
    'current_price': '0.004234'}],
  'originating_call_price': {'number_type': 'local',
   'base_price': None,
   'current_price': '0.002716'}},
 {'number': '+16288772320',
  'country': 'United States',
  'price_unit': 'GBP',
  'terminating_prefix_prices': [{'desti

In [ ]:
from services.knowledge_base.web_scrape import scrape_url

urls = ["https://api.slack.com/methods",
"https://api.slack.com/methods/conversations.replies",
"https://api.slack.com/methods/bots.info"]

# "https://api.slack.com/methods/dnd.teamInfo",
# "https://api.slack.com/methods/workflows.stepFailed",
# "https://api.slack.com/methods/search.all",
# "https://api.slack.com/methods/rtm.start",
# "https://api.slack.com/methods/functions.completeError",
# "https://api.slack.com/methods/chat.deleteScheduledMessage",
# "https://api.slack.com/methods/files.getUploadURLExternal",
# "https://api.slack.com/methods/reminders.list",
# "https://api.slack.com/methods/usergroups.update",
# "https://api.slack.com/methods/users.setActive",
# "https://api.slack.com/methods/stars.remove",
# "https://api.slack.com/methods/reactions.remove",
# "https://api.slack.com/methods/bookmarks.remove",
# "https://api.slack.com/methods/views.publish",
# "https://api.slack.com/methods/emoji.list",
# "https://api.slack.com/methods/team.info",
# "https://api.slack.com/methods/apps.uninstall",
# "https://api.slack.com/methods/migration.exchange",
# "https://api.slack.com/methods/dialog.open",
# "https://api.slack.com/methods/calls.info",
# "https://api.slack.com/methods/oauth.access",
# "https://api.slack.com/methods/pins.list",
# "https://api.slack.com/methods/apps.datastore.bulkGet",
# "https://api.slack.com/methods/admin.conversations.delete",
# "https://api.slack.com/methods/admin.users.setExpiration",
# "https://api.slack.com/methods/usergroups.users.update",
# "https://api.slack.com/methods/admin.roles.listAssignments",
# "https://api.slack.com/methods/admin.emoji.list",
# "https://api.slack.com/methods/admin.analytics.getFile",
# "https://api.slack.com/methods/admin.inviteRequests.deny",
# "https://api.slack.com/methods/files.remote.share",
# "https://api.slack.com/methods/apps.manifest.export",
# "https://api.slack.com/methods/team.billing.info",
# "https://api.slack.com/methods/oauth.v2.access",
# "https://api.slack.com/methods/calls.participants.add",
# "https://api.slack.com/methods/apps.activities.list",
# "https://api.slack.com/methods/tooling.tokens.rotate",
# "https://api.slack.com/methods/team.preferences.list",
# "https://api.slack.com/methods/assistant.threads.setStatus",
# "https://api.slack.com/methods/admin.apps.uninstall",
# "https://api.slack.com/methods/openid.connect.userInfo",
# "https://api.slack.com/methods/admin.usergroups.removeChannels",
# "https://api.slack.com/methods/admin.workflows.search",
# "https://api.slack.com/methods/conversations.canvases.create",
# "https://api.slack.com/methods/admin.teams.list",
# "https://api.slack.com/methods/canvases.access.set",
# "https://api.slack.com/methods/conversations.externalInvitePermissions.set",
# "https://api.slack.com/methods/chat.scheduledMessages.list",
# "https://api.slack.com/methods/admin.barriers.list",
# "https://api.slack.com/methods/auth.teams.list",
# "https://api.slack.com/methods/apps.connections.open",
# "https://api.slack.com/methods/files.comments.delete",
# "https://api.slack.com/methods/team.externalTeams.list",
# "https://api.slack.com/methods/users.discoverableContacts.lookup",
# "https://api.slack.com/methods/users.profile.get",
# "https://api.slack.com/methods/canvases.sections.lookup",
# "https://api.slack.com/methods/admin.functions.list",
# "https://api.slack.com/methods/admin.inviteRequests.denied.list",
# "https://api.slack.com/methods/functions.distributions.permissions.list",
# "https://api.slack.com/methods/admin.teams.settings.setName",
# "https://api.slack.com/methods/admin.users.session.getSettings",
# "https://api.slack.com/methods/admin.apps.approved.list",
# "https://api.slack.com/methods/admin.conversations.restrictAccess.listGroups",
# "https://api.slack.com/methods/admin.apps.config.lookup",
# "https://api.slack.com/methods/admin.teams.owners.list",
# "https://api.slack.com/methods/workflows.triggers.permissions.remove",
# "https://api.slack.com/methods/admin.auth.policy.removeEntities",
# "https://api.slack.com/methods/admin.functions.permissions.set",
# "https://api.slack.com/methods/admin.users.unsupportedVersions.export",
# "https://api.slack.com/methods/apps.auth.external.delete",
# "https://api.slack.com/methods/admin.conversations.ekm.listOriginalConnectedChannelInfo",
# "https://api.slack.com/methods/admin.workflows.collaborators.add",
# "https://api.slack.com/methods/conversations.requestSharedInvite.approve/test",
# "https://api.slack.com/methods/admin.apps.activities.list",
# "https://api.slack.com/methods/admin.workflows.permissions.lookup",
# "https://api.slack.com/methods/functions.workflows.steps.list",
# "https://api.slack.com/methods/admin.apps.restricted.list",
# "https://api.slack.com/methods/admin.teams.admins.list",
# "https://api.slack.com/methods/apps.event.authorizations.list",
# "https://api.slack.com/methods/admin.apps.requests.list",
# "https://api.slack.com/methods/admin.inviteRequests.approved.list",
# "https://api.slack.com/methods/functions.workflows.steps.responses.export",
# "https://api.slack.com/methods/admin.audit.anomaly.allow.updateItem",
# "https://api.slack.com/methods/canvases.delete",
# "https://api.slack.com/methods/api.test",
# "https://api.slack.com/methods/auth.revoke",
# "https://api.slack.com/methods/admin.workflows.triggers.types.permissions.set"]

await scrape_url(urls, "user_2mmXezcGmjZCf88gT2v2waCBsXv")



## Notion - 'object': 'error', 'status': 404, 'code': 'object_not_found',

In [ ]:
import requests

url = 'https://api.notion.com/v1/users'

NOTION_API_KEY = "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"

headers = {
    'Authorization': f'Bearer {NOTION_API_KEY}',
    'Notion-Version': '2022-06-28'
}

response = requests.get(url, headers=headers)
print(response.json())

In [17]:
from notion_client import Client
import os
from pprint import pprint

NOTION_API_KEY = "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"
PAGE_ID = "368c880f710a417582d411c17fc97df6"

# Initialize the Notion client
notion = Client(auth=NOTION_API_KEY)

def get_page_content(page_id) -> dict:
    try:
        # Fetch the page
        page = notion.pages.retrieve(page_id)
        
        # Fetch the page's content (blocks)
        blocks = notion.blocks.children.list(page_id)
        
        return {
            "page": page,
            "content": blocks["results"]
        }
    except Exception as e:
        print(f"Error fetching page content: {str(e)}")
        return None
    
def extract_block_text(block):
    """Extract plain text from a Notion block regardless of type"""
    block_type = block["type"]
    content = block[block_type]  # Get the content based on block type (paragraph, to_do, etc.)
    
    # If there's no rich_text or it's empty, return empty string
    if not content.get("rich_text"):
        return ""
    
    # Extract plain_text from the first rich_text element
    return content["rich_text"][0]["plain_text"]

page_content = get_page_content(PAGE_ID)

full_content = []
for block in page_content['content']:
    text = extract_block_text(block)
    if text:  # Only add non-empty strings
        full_content.append(text)

print("\n".join(full_content))  # Print all content together



#### Data structure to chunk and embed
- plain text of every block, noting it's type, 

In [ ]:
page_content

In [ ]:
""" CREATE NEW NOTION PAGE """

# url = 'https://api.notion.com/v1/blocks/9ee421b1e59f42b28e06a3089fc6f6c6/children?page_size=100'

url = 'https://api.notion.com/v1/pages'

NOTION_API_KEY = "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"

headers = {
    'Authorization': f'Bearer {NOTION_API_KEY}',
    'Notion-Version': '2022-06-28',
    'Content-Type': 'application/json'
}

data = {
	"parent": { "page_id": "4af64a52b6344443b9d7d84deef0e9df" },
	"properties": {
		"title": {
      "title": [{ "type": "text", "text": { "content": "A note from your pals at Notion" } }]
		}
	},
	"children": [
    {
      "object": "block",
      "type": "paragraph",
      "paragraph": {
        "rich_text": [{ "type": "text", "text": { "content": "You made this page using the Notion API. Pretty cool, huh? We hope you enjoy building with us." } }]
      }
    }
  ]
}

response = requests.post(url, headers=headers, json=data)
print(response.json())



In [ ]:
response.json()

### Why isn't all kb headers being returned?


In [ ]:
from services.knowledge_base.kb import get_kb_headers

headers = await get_kb_headers(current_user="user_2mmXezcGmjZCf88gT2v2waCBsXv")

headers

In [ ]:
headers

In [ ]:
import pandas as pd
import numpy as np
import json
from services.db import supabase_services

supabase = supabase_services.supabase_client()

# Read CSV file
df = pd.read_csv('stock_chunks_rows.csv')

# Clean array-like strings
def clean_array_data(value):
    if pd.isna(value):
        return None
    try:
        # If it's already a string representation of a list
        if isinstance(value, str):
            if value.startswith('[') and value.endswith(']'):
                # Remove the brackets and split by comma
                cleaned = value.strip('[]').replace('"', '').split(',')
                return [item.strip() for item in cleaned]
            return [value]
        elif isinstance(value, list):
            return value
        return [str(value)]
    except:
        return [str(value)]

# Clean the array columns
array_columns = ['heading', 'content']  # Add any other array columns here
for col in array_columns:
    if col in df.columns:
        df[col] = df[col].apply(clean_array_data)

# Replace NaN values with None
df = df.replace({np.nan: None})

# Process in smaller batches
batch_size = 500  # Reduced batch size for more reliable processing

def insert_batch(batch_df, batch_num):
    try:
        records = batch_df.to_dict('records')
        result = supabase.table('stock_chunks').insert(records).execute()
        print(f"Successfully inserted batch {batch_num} with {len(records)} records")
    except Exception as e:
        print(f"Error inserting batch {batch_num}: {str(e)}")
        if records:
            print(f"Sample record: {records[0]}")

# Process in batches
for i in range(0, len(df), batch_size):
    batch_df = df[i:i + batch_size]
    insert_batch(batch_df, i // batch_size + 1)

In [ ]:
from services.chat.gym_shark import chat_process

system_prompt = """
# Gym Shark AI Shopping Assistant System Prompt

You are an AI shopping assistant for Gym Shark, a popular fitness apparel and accessories brand. Your primary role is to help customers find the perfect products, answer their questions, and provide personalized recommendations. Always maintain a friendly, energetic, and supportive tone that aligns with Gym Shark's brand image of empowering athletes and fitness enthusiasts.

## Key Responsibilities:

1. Product Recommendations:
   - Based on customer preferences, workout types, body types, and style choices, suggest appropriate Gym Shark products.
   - Consider factors such as fabric, fit, color, and functionality when making recommendations.

2. Size and Fit Guidance:
   - Assist customers in finding the right size by asking about their measurements and preferences for fit (tight, loose, etc.).
   - Provide information on how different product lines may fit differently.

3. Product Information:
   - Offer detailed information about product features, materials, care instructions, and benefits.
   - Explain the technology behind Gym Shark's innovative fabrics and designs.

4. Outfit Coordination:
   - Help customers create complete outfits by suggesting complementary items.
   - Recommend products that work well for specific workout types or fitness goals.

5. Order and Shipping Information:
   - Provide information on ordering processes, shipping options, and estimated delivery times.
   - Assist with tracking orders and addressing any shipping-related concerns.

6. Returns and Exchanges:
   - Explain Gym Shark's return and exchange policies.
   - Guide customers through the process if they need to return or exchange an item.

7. Sales and Promotions:
   - Inform customers about ongoing sales, promotions, or special offers.
   - Suggest products that offer good value or are currently discounted.

8. Workout and Fitness Advice:
   - Offer basic workout tips and suggestions related to the products customers are interested in.
   - Provide general fitness motivation and encouragement.

9. Brand Information:
   - Share information about Gym Shark's history, mission, and values when relevant.
   - Highlight Gym Shark's commitment to sustainability and ethical practices.

## Guidelines:

- Always prioritize customer satisfaction and aim to understand their specific needs.
- Use positive, motivating language that encourages customers in their fitness journey.
- Be knowledgeable about fitness trends and how Gym Shark products align with them.
- If unsure about any product details or policies, advise the customer to check the official website or contact customer service.
- Respect customer privacy and never ask for personal information beyond what's necessary for product recommendations.
- Be prepared to handle common customer service scenarios with patience and professionalism.
- Stay updated on the latest Gym Shark product releases and collections.
- Use emojis sparingly to maintain a friendly yet professional tone.

Remember, your goal is to create a positive, helpful, and engaging shopping experience that reflects Gym Shark's commitment to quality, innovation, and customer satisfaction.

"""

await chat_process(user_message="hello", system_prompt=system_prompt, user_id="user_2lKpUPRJD4g5IErIdhbO7rBMn3K")



In [ ]:
from services.chat.lk_chat import lk_chat_process


lk_chat_process()

In [ ]:
async def main():
    # Call the function to get the generator
    async for item in lk_chat_process(message="hello", agent_id="a93ef199-3f74-4ab0-ac62-6284526d33d0"):
        print(item)

# For Jupyter notebook
await main()

In [ ]:
import anthropic
import base64
from IPython.display import Image, display

# Set your API key
anthropic.api_key = 'your_api_key_here'

In [ ]:
def encode_image(image_path):
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode('utf-8')

In [ ]:
import openai
from termcolor import colored
from dotenv import load_dotenv
import json
import os
import spacy
import requests
from supabase import create_client, Client
from openai import OpenAI
from tiktoken import encoding_for_model

load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")
SUPABASE_ANON_KEY = os.getenv("SUPABASE_ANON_KEY")

supabase: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)

openai = OpenAI()

In [ ]:
from services.chat.chat import similarity_search

results = await similarity_search("Accounts payable automation?", {'web': ['https://piqnic.com/'], 'text_files': []})

results


# chunking & embedding kb

in post.knowledge_base endpoint. when new item is added. we process that item through the below code.

In [ ]:
current_user = "user_2lKpUPRJD4g5IErIdhbO7rBMn3K"
items = supabase.table('knowledge_base').select('*').eq('user_id', current_user).execute()

In [ ]:

nlp = spacy.load("en_core_web_md")
data = items.data[1]['content']
doc = nlp(data)
cleaned_text = ' '.join([token.text for token in doc if not token.is_space and not token.is_punct])


def count_tokens(text, model="gpt-4o"):
    encoder = encoding_for_model(model)
    tokens = encoder.encode(text)
    return len(tokens)

def sliding_window_chunking(text, max_window_size=600, overlap=200):
    encoder = encoding_for_model("gpt-4o")  # Use the same model as in count_tokens
    tokens = encoder.encode(text)
    chunks = []
    start = 0
    while start < len(tokens):
        end = start + max_window_size
        chunk_tokens = tokens[start:end]
        chunk = encoder.decode(chunk_tokens)
        chunks.append(chunk)
        start += max_window_size - overlap
    return chunks

def insert_chunk(parent_id, content, chunk_index, embedding):
    print("func insert_chunk...")
    supabase.table('chunks').insert({
        'parent_id': parent_id,
        'content': content,
        'chunk_index': chunk_index,
        'embedding': embedding
    }).execute()

def get_embedding(text):
    response = openai.embeddings.create(
        input=text,
        model="text-embedding-3-small"
    )
    return response.data[0].embedding

def process_item(item_id, content):
    print("func process_item...")
    chunks = sliding_window_chunking(content) 
    for index, chunk in enumerate(chunks):
        embedding = get_embedding(chunk)
        print("index", index)
        print("chunk", chunk)
        print("embedding", embedding)
        insert_chunk(item_id, chunk, index, embedding)
#process_item(item_id=items.data[1]['id'], content=cleaned_text)



### Rerank RAG

In [ ]:
def similarity_search(query, table_name, match_threshold=0.2, match_count=10):
    query_embedding = get_embedding(query)
    
    response = supabase.rpc(
        'match_documents',
        {
            'query_embedding': query_embedding,
            'match_threshold': match_threshold,
            'match_count': match_count,
            'table_name': table_name
        }
    ).execute()
    return response.data


def rerank_documents(user_query, top_n, docs):
    url = 'https://api.jina.ai/v1/rerank'
    headers = {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer jina_b716ce28cd1b49bc920e57a5bfb6de061z36vM3vogg6y-_2d5qcoXHe_rdo'
    }
    data = {
        "model": "jina-reranker-v2-base-multilingual",
        "query": user_query,
        "top_n": top_n,
        "documents": docs
    }
    response = requests.post(url, headers=headers, json=data)
    reranked_docs = response.json()['results']
    reranked_docs = [i['document']['text'] for i in reranked_docs]
    return reranked_docs


def rag_response(user_query):
    table_name = "chunks"
    results = similarity_search(user_query,table_name)
    docs = [result['content'] for result in results]
    reranked_docs = rerank_documents(user_query, 3, docs)

    return reranked_docs

In [ ]:
system_prompt = """
You are a helpful assistant designed to search the company knowledge base, and find relevant information to answer questions from users.

Be conversational and friendly, and at times informal, while maintaining a dignified professional persona.

Where a question from the user appears to be best answered by information from the knowledge base, you will use the <context> to augment your response to the user.

Where your responses involved listing, or providing of information. Format them in markdown to allow for pretty displaying to the user to enable intuitive and quick understanding of the information you have kindly provided.

"""
conversation_history = {
    "user_history": [],
    "assistant_history": [],
    "function_history": []
}

def llm_response(system_prompt, user_prompt, conversation_history):
    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt},
    ]

    response = openai.chat.completions.create(
        model="gpt-4o",
        messages=messages,
        stream=False
    )

    # full_response = ""
    # tool_calls = []

    # for chunk in response:
    #     delta = chunk.choices[0].delta
    #     if delta.content:
    #         yield delta.content
    #         full_response += delta.content
    #     if delta.tool_calls:
    #         tool_calls.extend(delta.tool_calls)

    # conversation_history["user_history"].append({"role": 'user', "content": user_prompt})
    # conversation_history["assistant_history"].append({"role": 'assistant', "content": response})
    return response



In [ ]:

user_query = "tell me a little about flowon"
retrieved_docs = rag_response(user_query)
user_prompt = f"""{user_query}
retrieved docs {retrieved_docs} """

response = llm_response(system_prompt, user_prompt, conversation_history)

# response_received = False
# for response_chunk in llm_response(system_prompt, user_prompt, conversation_history):
#     response_received = True
#     print(response_chunk, end='', flush=True)

response.choices[0].message.content